### DISCUSSION TITLE
x-if seems buggy ('sticky')?

### DISCUSSION THREAD

**Discussion by tbjolset** [alpinejs/alpine#2540] [7. Bugs (archived)]
x-if seems to be a nice way to do component based design: use x-if to only evaluate/render a block if the element exists, otherwise ignore the object/component. However, there seems to be some weird behavior that this works initially if the object in question is null, but if you set the object non-null, then back to null, the x-if template insists on trying to evaluate the object. Easier to see from an example:
```
<head>
  <script src="//unpkg.com/alpinejs" defer></script>
</head>

<body x-data="{ user: null }">
  <template x-if="user">
    <div>
      <div x-text="'user: ' + user.name"></div>
      <button @click="user = null">Log out</button>
    </div>
  </template>
  <template x-if="!user">
    <button @click="user = { name: 'lebowski' }">Log in</button>
  </template>

</body>
```
Initially this works perfectly: The x-if template doesn't try to render/evaluate the logged in block, since the user is null. When tapping log in, it also shows the username correctly. However, after tapping log out, the x-if template now renders the log in div again, as it should, but also tries to evaluate user.name, for some reason, leading to uncaught type errors. 
`alpinejs:1 Alpine Expression Error: Cannot read properties of null (reading 'name')`

**Comment by kylethielk** [comment]
hey @tbjolset unfortunately this is a bug. Its been reported a few other times as well see #2296 #1837 #1917

Essentially `x-if` does not currently guarantee that any child expressions will not run when `x-if` evaluates to false.

**Reply by kylethielk** [reply]
Forgot to mention, the way to work-around this would be with [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) but you lose a ton of safari users with it as only very recent versions support it.

```
  <template x-if="user">
    <div>
      <div x-text="'user: ' + user?.name"></div>
      <button @click="user = null">Log out</button>
    </div>
  </template>
```
Alternatively just use a check like so:
```
  <template x-if="user">
    <div>
      <div x-text="'user: ' + (!!user ? user.name : '')"></div>
      <button @click="user = null">Log out</button>
    </div>
  </template>
```

**Reply by SimoTod** [reply]
@kylethielk I believe you can also use `<div x-text="'user: ' + (user && user.name)"></div>` which should be supported by earlier versions of Safari although it's probably not immediate to understand.

**Reply by tbjolset** [reply]
Thanks for the response @kylethielk  and @SimoTod. i know the workarounds but avoiding that boilerplate code was the whole reason for using x-if in the first place. I cannot see any reason to use x-if instead of x-show now? I still need to manually check the object everywhere inside the template, and in addition i have the weird behavior of acing differently on init vs after toggling - as well as losing the ability to do transitions.
In my head right now, it would be better to remove x-if completely, as it offers no advantages over x-show and causes hard to understand bugs?

**Reply by SimoTod** [reply]
For small portions of html x-show is probably faster as it doesn't attach and remove nodes continuously but i suspect you'll need to check if user exists anyway.

It seems a bit extreme to remove a feature because there is a bug, though. the project is open source so someone will fix it eventually. Have you had a look to understand where the issue is?

**Reply by kylethielk** [reply]
I spent some time looking through the inner workings of Alpine and submitted a PR that definitely solves the provided example, but also (I believe) solves the more general issue of sub expressions being evaluated after x-if evaluates to false: https://github.com/alpinejs/alpine/pull/2556

**Reply by SimoTod** [reply]
@kylethielk yeah, I suspected that the issue was related to the fact that the effect was already queued. 

To be fair, I think it's a bug (or at least a weird design decision) in Vue reactivity (https://github.com/vuejs/vue-next/blob/v3.1.5/packages/reactivity/src/effect.ts#L102) combined with the fact that Alpine should probably call destroyTree in the cleanUp callback rather than relying on the mutation observer to pick it up (wrapped in a `mutateDom` call).
Basically, when an effect is released, the active flag changes to false and, when it's time to run the effect, Vue checks that flag and execute the callback instead of skipping it which I find confusing.

Theoretically, the same issue happens with x-for if used in a non-canonical way.
```html
<div x-data="{ users: [{ name: 'lebowski' }] }">
  <template x-for="(user, idx) in users">
    <div x-text="'user: ' + users[idx].name"></div>
  </template>
  <button @click="users = []">Reset</button>
</div>
```

**Reply by kylethielk** [reply]
Great point @SimoTod. I've updated PR to include a fix for x-for case as well.

**Reply by janzikmund** [reply]
@kylethielk - thanks so much for https://github.com/alpinejs/alpine/pull/2556, it looks like for some reason the bug still shows up if there is `x-modelable` directive used within the `<template>`. 

Please see the attached sample code below, I also made a [fiddle](https://jsfiddle.net/janzikmund/Ld5wkzs7/)

```html
<html>
<head><script src="https://unpkg.com/alpinejs" defer></script></head>
<body>
    <div x-data="{ user: { number: 5 }}">
        <template x-if="user">
            <div>
                <div x-data="{ count: 0 }" x-modelable="count" x-model="user.number">
                    <button @click="count++">Increment</button>
                </div>
                Number: <span x-text="user.number"></span>
            </div>
        </template>
        <button @click="user=null">remove user</button>
    </div>
</body>
```

The weird thing is, that if you click the **remove user** button as first, you get the errors. While if you at least once click the **increment** and then click **remove user**, the errors don't show up anymore. Do you have any idea what's behind this?

**Reply by kylethielk** [reply]
Hi @janzikmund unfortunately `x-modelable` was released after I pushed my fix so I am not familiar with its inner workings. It might be worth looking to see if [my changes](https://github.com/alpinejs/alpine/pull/2556/files#) are applicable to the `x-modelable` directive: https://github.com/alpinejs/alpine/pull/2654/files#diff-af620dcff7cc59b1c7efe7e9cef800c8622ab637447075c1cd725d3d8ab8e306

**Reply by janzikmund** [reply]
@kylethielk - thanks, I checked your code, but I still wasn't able to pinpoint the problem. Apparently it's caused by the [`effect(() => innerSet(outerGet()))`](https://github.com/alpinejs/alpine/blob/4df4e4f12aa929e69cde0cd0fde65ac624b8e5d8/packages/alpinejs/src/directives/x-modelable.js#L25) line in `x-modelable` directive. If I remove it, the errors are not triggered anymore, but obviously the value from x-model isn't transformed.

Strange thing is, the error doesn't only show up for the element with `x-modelable`, but also for all the others in the same x-if; the single `x-modelable` is enough to start raising it for all other elements. See my [updated fiddle](https://jsfiddle.net/janzikmund/Ld5wkzs7/3/) - the error shows for all three fields - count, name and country.

I tried to check and fix it in `x-if` similarly to what you did, only the `dequeueJob` didn't seem to work and the effect still showed up in console, so I added a different code as per below. Now the _x_effects queue in console is empty, but still the errors show up the same, not sure where else they can be wired:

````javascript
el._x_undoIf = () => {
  walk(clone2, (node) => {
    if (!!node._x_effects) {
      node._x_effects.forEach(dequeueJob);

      // one way
      node._x_effects.forEach((effect) => {
        node._x_effects.delete(effect);
      });
      // another way, seems to do the same
      node._x_effects.clear(); 
      // shows empty _x_effects, but still the errors are thrown
      console.log(node, node._x_effects);  
    }
  });
  clone2.remove();
  delete el._x_currentIfEl;
};
````

Also, this _x_effects clean up is only happening on the element with x-modelable, while the errors are thrown for all the other fields as well, so the problem must probably lie somewhere else.

**Reply by janzikmund** [reply]
For anyone interested, I haven't found a way to fix this in Alpinejs, but the easiest way to circumnavigate the problem was to bind the x-if on a different variable than the one assigned to x-model on the modelable field, then no problem seems to happen anymore. For my example above, the modified code would look something like this:

```` javascript
<html>
<head><script src="https://unpkg.com/alpinejs" defer></script></head>
<body>
    <div x-data="{ userVisible: true, user: { number: 5 }}">
        <template x-if="userVisible">
            <div>
                <div x-data="{ count: 0 }" x-modelable="count" x-model="user.number">
                    <button @click="count++">Increment</button>
                </div>
                Number: <span x-text="user.number"></span>
            </div>
        </template>
        <button @click="userVisible=false;user=null;">remove user</button>
    </div>
</body>
````

**Comment by Rock1raw** [comment]
V

**Reply by ekwoka** [reply]
If you bumped this for having the same issue, it was fixed already.

**Reply by stusherwin** [reply]
@ekwoka it still seems to be broken? I'm having the same issue with `x-if` in v3.15.3. From looking at the code, it was fixed in https://github.com/alpinejs/alpine/pull/2556 - this line is for the x-if directive:
https://github.com/alpinejs/alpine/commit/d49a0b0b5841d05e41b70dc806d80c2208765666#diff-aeb86af77b35b0c71e72f5f0c0ebafce1ce4c2e6a026ef229af7797e7802126eR28

But it looks as if this was reverted in https://github.com/alpinejs/alpine/pull/4300:
https://github.com/alpinejs/alpine/commit/6ac97825351623e9251b6a8bb7cbd805b7a9a94c#diff-aeb86af77b35b0c71e72f5f0c0ebafce1ce4c2e6a026ef229af7797e7802126eL33

I can't see anything that calls `dequeueJob()` on the existing side effects any more, and it seems to me that was the fix?

**Reply by ekwoka** [reply]
Yeah, so the second one is supposed to fix it by instead of trying to dequeueJobs, its supposed to synchronously clean the entire tree.

The dequeueJob method failed at this due to how the tree was essentially disconnected causing the tree not to be dequeued properly in most cases.

I'll check again, but that was supposed to fix it (and did at the time it was implemented).

**Reply by stusherwin** [reply]
Hi @ekwoka thanks for checking, but I figured out what the issue was and it wasn't anything to do with this bug.

I had an `x-if` with an `x-show` within it:
```
<template x-if="pageState.page == 'customers'">
  <div>
    <h1>Customers</h1>
    <div x-show="!pageState.data.ready">Loading...</div>
    <div x-show="pageState.data.ready">Loaded</div>  
  </div>
</template>
```

And then on an event handler I was doing (simplified):
```
// clean up old page
pageState.data.ready = false;

// set new page
pageState = {
  page: 'products'
};
```

The x-show was trying to respond to the change in state of `pageState.data.ready` but I was resetting the `pageState` object before the effect handler got round to executing. I fixed it by adding a timeout:
```
// clean up old page
pageState.data.ready = false;

// set new page
setTimeout(() => {
  pageState = {
    page: 'products'
  };
});
```

**Reply by ekwoka** [reply]
That makes some more sense. I'd probably recommend ensuring you always maintain the data structure, or just use `?.` optional chaining.
